### PR TITLE
feat(api): implement /users endpoint

### PR DESCRIPTION
**Pull request by asastre**
## Description
- Implementation of users routes for CRUD operations. No authentication is done so far.
- Added validations to the parameters received in the requests using `express-validator`.
- Added missing method `deleteUser` in the `DataProvider` interface and its implementations.
- Add validation in patch and put routes to ensure the ID received in the param is the same as the one received in the body.
- Refactor validations to unify common validations for events and users endpoints.

## Context
The api endpoint of users were not implemented yet.

## Approach taken / Explain the design
The same way it has been done with events endpoint, all the data validation has been done based on the user type declaration.
For the `patch` request, in order to take advantage of all the validantions done, a method has been implemented to transform all fields to optional fields but still keep all the other validations.

## Testing

The pull request has no tests.

### CODE REVIEW

**Review by vanbasten17 — APPROVED** [review]

File: packages/botonic-api/src/rest/routes/events.ts [hunk 1]
```diff
@@ -109,26 +102,26 @@ router
           .send({ error: `Event with ID '${params.eventId}' not found` })
         return
       }
+      if (event.eventId !== updatedEvent.eventId) {
+        res
+          .status(400)
+          .send({ error: `Both event ID (params and body) must match` })
+        return
+      }
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/events.ts:110]: This shouldn't be needed. The ID is already encoded in the URL, there should be no ID in the body (or if there is, just ignore it and always use the URL one)

File: packages/botonic-api/src/rest/routes/events.ts [hunk 2]
```diff
@@ -82,20 +81,14 @@ router
       res.status(500).send({ error: e.message })
     }
   })
-  .put(checkSchema({ eventId: eventIdParamSchema }), async (req, res) => {
-    const errors = await validateBotonicEventData(req)
+  .put(async (req, res) => {
+    const errors = await validateBotonicEventData(req, false, true)
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/events.ts:85]: named parameters better? it's hard to understand what this does without looking at the function signature.

File: packages/botonic-api/src/rest/routes/events.ts [hunk 3]
```diff
@@ -142,6 +135,12 @@ router
           .send({ error: `Event with ID '${params.eventId}' not found` })
         return
       }
+      if (newEventData.eventId && event.eventId !== newEventData.eventId) {
+        res
+          .status(400)
+          .send({ error: `Both user ID (params and body) must match` })
+        return
+      }
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/events.ts:143]: Same as with User

File: packages/botonic-api/src/rest/routes/users.ts [hunk 4]
```diff
@@ -1,55 +1,182 @@
...
   .route('/')
-  .get((_, res) => {
-    res.send('GET HTTP method on user resource')
-  })
-  .post((_, res) => {
-    res.send('POST HTTP method on user resource')
+  .get(
+    checkSchema({
+      limit: limitParamSchema,
+      offset: offsetParamSchema,
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/users.ts:24]: The most common way to add pagination to an API is to define a param "page" (by default 1) and an optional "pageSize" (default typically 10 or 20), with pageSize having a maximum allowed (for instance 100).
Is there any particular reason you used "limit" and "offset"?

File: packages/botonic-api/src/rest/routes/users.ts [hunk 5]
```diff
@@ -1,55 +1,182 @@
...
+          .status(404)
+          .send({ error: `User with ID '${params.userId}' not found` })
+        return
+      }
+      if (user.id !== updatedUser.id) {
+        res
+          .status(400)
+          .send({ error: `Both user ID (params and body) must match` })
+        return
+      }
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/users.ts:115]: not needed

File: packages/botonic-api/src/rest/routes/users.ts [hunk 6]
```diff
@@ -1,55 +1,182 @@
...
+            .status(404)
+            .send({ error: `User with ID '${params.userId}' not found` })
+          return
+        }
+        if (newUserData.id && user.id !== newUserData.id) {
+          res
+            .status(400)
+            .send({ error: `Both user ID (params and body) must match` })
+          return
+        }
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/users.ts:150]: not needed

**Review by ericmarcos — CHANGES_REQUESTED** [review]

File: packages/botonic-api/src/rest/routes/users.ts [hunk 7]
```diff
@@ -1,55 +1,182 @@
...
+        const errors = validationResult(req)
+        if (!errors.isEmpty()) {
+          res.status(400).send({ errors: errors.array() })
+          return
+        }
+
+        const query = matchedData(req, { locations: ['query'] })
+        const dp = dataProviderFactory(process.env.DATA_PROVIDER_URL)
+        const users = await dp.getUsers(query.limit, query.offset)
+        res.status(200).send(users)
```
Comment (ericmarcos) [inline packages/botonic-api/src/rest/routes/users.ts:37]: Usually when returning results in a paginated API you'd expect an object like this:

```
{
  "prev": "<link to prev page>",
  "next": "<link to next page>",
  "count": <total number of records>,
  "results": [{...user1}, {...user2}, ...]
}
```

File: packages/botonic-api/src/rest/routes/events.ts (hunk 1 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/events.ts]: Make sense, fixed!

File: packages/botonic-api/src/rest/routes/events.ts (hunk 2 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/events.ts]: Agree, changed!

File: packages/botonic-api/src/rest/routes/events.ts (hunk 3 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/events.ts]: Removed!

File: packages/botonic-api/src/rest/routes/users.ts (hunk 5 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/users.ts]: Removed!

File: packages/botonic-api/src/rest/routes/users.ts (hunk 6 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/users.ts]: Removed!

File: packages/botonic-api/src/rest/routes/users.ts (hunk 4 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/users.ts:24]: I've seen it done both ways and decided to use the limit/offset one. 
For the developer maybe it's easier to use the page/pageSize...I'll change it to page/pageSize in another PR and I'll add also the `prev`, `next`, `count`, `results` in the api response as you said in another comment.

File: packages/botonic-api/src/rest/routes/users.ts (hunk 7 above)
Comment (asastre) [reply packages/botonic-api/src/rest/routes/users.ts:37]: :+1: I'll add this into another PR with the `page`/`pageSize` refactor.

**Review by ericmarcos — APPROVED** [review]

**Pull request merged**